src/hotspot/share/oops/instanceKlass.cpp

```diff
@@ -3739,10 +3740,13 @@ const char* InstanceKlass::internal_name() const {
 void InstanceKlass::print_class_load_logging(ClassLoaderData* loader_data,
                                              const ModuleEntry* module_entry,
                                              const ClassFileStream* cfs) const {
+
   if (ClassListWriter::is_enabled()) {
     ClassListWriter::write(this, cfs);
   }
 
+  print_class_load_cause_logging();
+
```

Comment:
This prints the stacks before printing the name of the class. I think the name should be printed first.

Reply:
Not sure what you mean. The `class+load+cause` log output includes the name before the stack. See the updated examples in this PR's description.

If you mean `class+load+cause` log output precedes `class+load` log output, then yes, that is the case and it's intentional to keep these logging features independent of each other as David suggested (https://github.com/openjdk/jdk/pull/14553#discussion_r1249847278). Since they are independent, it should not matter which output comes first.

Reply:
Even if they are independent, relevant output should be grouped together. Currently, if we specify `-Xlog:class+load*=debug`, the output would look like this:

```
Java stack when loading  java.util.concurrent.ConcurrentHashMap$ValuesView
...
[many many lines of output]
...
java.util.concurrent.ConcurrentHashMap$ValuesView source: shared objects file
  klass: 0x000000080033a0f8 super: 0x000000080033a320 loader: [loader data: 0x00007faa34113890 of 'bootstrap']
```
This makes it difficult to see all the information of this class at a glance.

Reply:
If I'm interested in class load cause logging of `java.util.concurrent.ConcurrentHashMap$ValuesView`, I don't particularly care about the `class+load` output for that class.
Note that some of the `class+load+placeholders` output is going to be shown after `class+load+cause` anyway.
That said, I've pushed a commit to implement your request: d84186840b874ba74e270f6ac51cc0ed71d655f5

Reply:
It's not about personal preferences. We should organize the logging so that they are grouped logically. That way, they can be more useful for others who might use the log tags in a different way.

The `class+load+placeholders` logs are for dealing with parallel loading of the same class by different threads. This happens before the class is actually loaded, so those logs cannot appear after the main `class+load` tag.

***

To minimize the delta in this PR, and limit code indentation, it's better to move the old code to a new function, so you have something like:

```
void InstanceKlass::print_class_load_logging(ClassLoaderData* loader_data,
                                             const ModuleEntry* module_entry,
                                             const ClassFileStream* cfs) const {

  if (ClassListWriter::is_enabled()) {
    ClassListWriter::write(this, cfs);
  }

>>> add
  print_class_load_helper();
  print_class_load_cause_logging();
}

void InstanceKlass::print_class_load_helper(ClassLoaderData* loader_data,
                                             const ModuleEntry* module_entry,
                                             const ClassFileStream* cfs) const {
<<< end
```

Reply:
> We should organize the logging so that they are grouped logically.

There's nothing ungrouped about the original order; the order of `class+load` and `class+load+cause` were swapped but they were still grouped together.

> This happens before the class is actually loaded, so those logs cannot appear after the main class+load tag.

Maybe I'm interpreting the output incorrectly, but that's not what I see:
```
> java "-Xlog:class+load=debug,class+load+cause=info,class+load+placeholders=debug" -XX:LogClassLoadingCauseFor=java.lang.StringCoding --version
...
[1.917s][debug][class,load,placeholders] entry java/lang/StringCoding : find_and_add LOAD_INSTANCE
[1.917s][debug][class,load,placeholders] loadInstanceThreadQ threads:0x000000013080a000,
[1.917s][debug][class,load,placeholders] superThreadQ threads:
[1.917s][debug][class,load,placeholders] defineThreadQ threads:
[1.917s][info ][class,load             ] java.lang.StringCoding source: /Users/dnsimon/dev/jdk-jdk/open/build/macosx-aarch64/jdk/modules/java.base
[1.917s][debug][class,load             ]  klass: 0x000000030007acf8 super: 0x0000000300041170 loader: [loader data: 0x00006000034ccaa0 of 'bootstrap'] bytes: 1361 checksum: af8c723a
[1.917s][info ][class,load,cause       ] Java stack when loading java.lang.StringCoding:
[1.917s][info ][class,load,cause       ]        at java.lang.String.encodeUTF8(java.base/String.java:1302)
[1.917s][info ][class,load,cause       ]        at java.lang.String.encode(java.base/String.java:867)
[1.917s][info ][class,load,cause       ]        at java.lang.String.getBytes(java.base/String.java:1818)
[1.917s][info ][class,load,cause       ]        at sun.nio.fs.Util.toBytes(java.base/Util.java:55)
[1.917s][info ][class,load,cause       ]        at sun.nio.fs.UnixFileSystem.<init>(java.base/UnixFileSystem.java:82)
[1.917s][info ][class,load,cause       ]        at sun.nio.fs.BsdFileSystem.<init>(java.base/BsdFileSystem.java:50)
[1.917s][info ][class,load,cause       ]        at sun.nio.fs.MacOSXFileSystem.<init>(java.base/MacOSXFileSystem.java:52)
[1.917s][info ][class,load,cause       ]        at sun.nio.fs.MacOSXFileSystemProvider.newFileSystem(java.base/MacOSXFileSystemProvider.java:44)
[1.917s][info ][class,load,cause       ]        at sun.nio.fs.MacOSXFileSystemProvider.newFileSystem(java.base/MacOSXFileSystemProvider.java:37)
[1.917s][info ][class,load,cause       ]        at sun.nio.fs.UnixFileSystemProvider.<init>(java.base/UnixFileSystemProvider.java:78)
[1.917s][info ][class,load,cause       ]        at sun.nio.fs.BsdFileSystemProvider.<init>(java.base/BsdFileSystemProvider.java:38)
[1.917s][info ][class,load,cause       ]        at sun.nio.fs.MacOSXFileSystemProvider.<init>(java.base/MacOSXFileSystemProvider.java:39)
[1.917s][info ][class,load,cause       ]        at sun.nio.fs.DefaultFileSystemProvider.<clinit>(java.base/DefaultFileSystemProvider.java:35)
[1.917s][info ][class,load,cause       ]        at java.nio.file.FileSystems.getDefault(java.base/FileSystems.java:186)
[1.917s][info ][class,load,cause       ]        at java.nio.file.Path.of(java.base/Path.java:148)
[1.917s][info ][class,load,cause       ]        at jdk.internal.module.SystemModuleFinders.ofSystem(java.base/SystemModuleFinders.java:188)
[1.917s][info ][class,load,cause       ]        at jdk.internal.module.ModuleBootstrap.boot2(java.base/ModuleBootstrap.java:244)
[1.917s][info ][class,load,cause       ]        at jdk.internal.module.ModuleBootstrap.boot(java.base/ModuleBootstrap.java:174)
[1.917s][info ][class,load,cause       ]        at java.lang.System.initPhase2(java.base/System.java:2227)
[1.917s][debug][class,load,placeholders] entry java/lang/StringCoding : find_and_add DEFINE_CLASS
[1.917s][debug][class,load,placeholders] loadInstanceThreadQ threads:0x000000013080a000,
[1.917s][debug][class,load,placeholders] superThreadQ threads:
[1.917s][debug][class,load,placeholders] defineThreadQ threads:0x000000013080a000,
[1.917s][debug][class,load,placeholders] entry java/lang/StringCoding : find_and_remove DEFINE_CLASS , InstanceKlass 'java/lang/StringCoding'
[1.917s][debug][class,load,placeholders] loadInstanceThreadQ threads:0x000000013080a000,
[1.917s][debug][class,load,placeholders] superThreadQ threads:
[1.917s][debug][class,load,placeholders] defineThreadQ threads:0x000000013080a000,
[1.917s][debug][class,load,placeholders] entry java/lang/StringCoding : find_and_remove LOAD_INSTANCE , InstanceKlass 'java/lang/StringCoding'
...
```

I've pushed another commit with your suggested change: 4b910b16dbd6f72a3d74fc0c72703aec5e7d9671

Reply:
> > We should organize the logging so that they are grouped logically.
> 
> There's nothing ungrouped about the original order; the order of `class+load` and `class+load+cause` were swapped but they were still grouped together.

My point is if you want to see both the stack and the loader for a class, now you can see them right next to each other in the log file, without having to scroll through pages of call stacks.

The new version looks good to me.

> > This happens before the class is actually loaded, so those logs cannot appear after the main class+load tag.
> 
> Maybe I'm interpreting the output incorrectly, but that's not what I see:
> 
> ```
> > java "-Xlog:class+load=debug,class+load+cause=info,class+load+placeholders=debug" -XX:LogClassLoadingCauseFor=java.lang.StringCoding --version
> ...
> [1.917s][debug][class,load,placeholders] entry java/lang/StringCoding : find_and_add LOAD_INSTANCE
> [1.917s][debug][class,load,placeholders] loadInstanceThreadQ threads:0x000000013080a000,
> [1.917s][debug][class,load,placeholders] superThreadQ threads:
> [1.917s][debug][class,load,placeholders] defineThreadQ threads:
> [1.917s][info ][class,load             ] java.lang.StringCoding source: /Users/dnsimon/dev/jdk-jdk/open/build/macosx-aarch64/jdk/modules/java.base
> [1.917s][debug][class,load             ]  klass: 0x000000030007acf8 super: 0x0000000300041170 loader: [loader data: 0x00006000034ccaa0 of 'bootstrap'] bytes: 1361 checksum: af8c723a
> [1.917s][info ][class,load,cause       ] Java stack when loading java.lang.StringCoding:

Before the class is actually parsed and loaded, its name is first entered into the place holder table. The logging reflects that.